examples/twittermap/web/public/javascripts/map/countmap.js

```diff
@@ -2,6 +2,208 @@ angular.module('cloudberry.map')
+        diffArray[newArray[i]] = true;
+      }
+      for (var j = 0; j < oldArray.length; j++) {
+        if (diffArray[oldArray[j]]) {
+          delete diffArray[oldArray[j]];
+        } else {
+          diffArray[oldArray[j]] = true;
+        }
+      }
+      for (var key in diffArray) {
+        difference.push(key);
+      }
+      return difference;
+    }
+
+    // Concat two hashmap results
+    function concatHashmap(newMap, cachedMap) {
+      if (cachedMap.count()===0) {
+        return newMap;
+      }
+
+      var concatMap = new HashMap();
+      newMap.forEach(function(value, key){
+        var concatValue = [];
+        var cacheValue = cachedMap.get(key);
+        if(value !== 0) {
+          if (cacheValue === undefined) {
+            concatValue=value;
+          }
+          else {
+            concatValue = value.concat(cacheValue);
+          }
+        }
+        else if (cacheValue !== undefined) {
+          concatValue=cacheValue;
+        }
+
+        concatMap.set(key,concatValue);
+      });
+      return concatMap;
+    }
+
+    // sum of one element in an array of objects
+    function sum(items, prop){
+      return items.reduce( function(previousVal, currentVal){
+        return previousVal + currentVal[prop];
+      }, 0);
+    }
+
+    function getPopupContent() {
+      // get chart data for the polygon
+      var geoIDChartData = $scope.ChartDataMap.get($scope.selectedGeoID);
+      (geoIDChartData && geoIDChartData.length!==0)? $scope.chartData = $scope.preProcess(geoIDChartData) : $scope.chartData = [];
+
+      // get the count info of polygon
+      var placeName = $scope.selectedPlace.properties.name;
+      var infoPromp = $scope.infoPromp;
+      var logicLevel = $scope.status.logicLevel;
+      var count = sum($scope.chartData,"y");
+
+      // Generate the html in pop up window
+      var content;
+      if($scope.chartData.length===0) {
+        content = "<div id=\"popup-info\" style=\"margin-bottom: 0\">" +
+          "<div id=\"popup-statename\">"+logicLevel+": "+placeName+"</div>" +
+          "<div id=\"popup-count\" style=\"margin-bottom: 0\">"+infoPromp+"<b> "+count+"</b></div>" +
+          "</div>"+
+          "<canvas id=\"myChart\" height=\"0\" ></canvas>";
+      }else {
+        content = "<div id=\"popup-info\">" +
+          "<div id=\"popup-statename\">"+logicLevel+": "+placeName+"</div>" +
+          "<div id=\"popup-count\">"+infoPromp+"<b> "+count+"</b></div>" +
+          "</div>"+
+          "<canvas id=\"myChart\"></canvas>";
+      }
+      return content;
+    }
+
+    // Add the event for popup window: when mouse out, close the popup window
+    function addPopupEvent() {
+      document.getElementsByClassName("leaflet-popup")[0].onmouseout=function (e) {
+        var target = e.relatedTarget;
+        if(target && (target.className.toString()==="[object SVGAnimatedString]"||target.className.toString().substring(0,4)==="form")) {
```

Comment:
The logic here are not that elegant and straightforward about checking target element class, could you add some comments above this line to remind future developers about what's happening here, and why we cannot `closePopup()` without any check once `mouseout` happens.

Reply:
Close popup when the mouse out of popup window and:
1. move into the area of map without polygons
2. Or move into the search bar

When the mouse move into the polygon which is the owner of popup window, it should not be close.
If we close it without checking, when the mouse move out from popup window, and move into its own polygon, the window will close and reopen, which means it will "flash". To avoid it, we need to check what is the element the mouse move in.